File: packages/scopecat/src/scopecat/records/scientific_selection.py
"""Scientific intent and the exact evidence retained by a checked preview."""

from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, JsonValue

from scopecat.records.config_context import ConfigContextRef, ContextRunConfigSource
from scopecat.records.parameter_revision import ParameterRevisionRef
from scopecat.records.parameter_update import ParameterUpdate
from scopecat.records.plan_ref import PlanConfigRef
from scopecat.records.run import (
    AnalysisCandidateRunConfigSource,
    ConfigRegistryRunConfigSource,
    ParameterRunConfigSource,
)
from scopecat.records.sample import SampleId
from scopecat.records.scientific_binding import (
    InlineSamplesSubject,
    RegisteredTargetSubject,
    ResolvedScientificBinding,
    UnboundSubject,
)
from scopecat.records.scientific_scope import BatchScope, DeclaredBatch, UnscopedBatch
from scopecat.records.setup import SetupRevisionRef
from scopecat.records.target_catalog import TargetRevisionRef


class _SelectionModel(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)


class UnboundSubjectChoice(_SelectionModel):
    kind: Literal["unbound"] = "unbound"


class SampleSubjectChoice(_SelectionModel):
    kind: Literal["sample"] = "sample"
    sample_id: SampleId
    revision: int | None = Field(default=None, ge=1)


class RegisteredTargetChoice(_SelectionModel):
    kind: Literal["registered_target"] = "registered_target"
    ref: TargetRevisionRef


type SubjectChoice = Annotated[
    UnboundSubjectChoice | SampleSubjectChoice | RegisteredTargetChoice,
    Field(discriminator="kind"),
]


class ActiveConfiguration(_SelectionModel):
    kind: Literal["active"] = "active"


class SavedConfiguration(_SelectionModel):
    kind: Literal["saved"] = "saved"
    ref: PlanConfigRef


class ParameterConfiguration(_SelectionModel):
    """Independent parameters; omitted setup resolves current authority at preview."""

    kind: Literal["parameters"] = "parameters"
    ref: ParameterRevisionRef
    setup: SetupRevisionRef | None = None
    overrides: tuple[ParameterUpdate, ...] = Field(default=(), max_length=256)


class WorkingPointConfiguration(_SelectionModel):
    kind: Literal["working_point"] = "working_point"
    ref: ConfigContextRef
    overrides: tuple[ParameterUpdate, ...] = Field(default=(), max_length=256)


class CandidateConfiguration(_SelectionModel):
    kind: Literal["candidate"] = "candidate"
    source: AnalysisCandidateRunConfigSource


type ConfigurationChoice = Annotated[
    ActiveConfiguration
    | ParameterConfiguration
    | SavedConfiguration
    | WorkingPointConfiguration
    | CandidateConfiguration,
    Field(discriminator="kind"),
]

type LaunchConfigSource = (
    ConfigRegistryRunConfigSource
    | ParameterRunConfigSource
    | ContextRunConfigSource
    | AnalysisCandidateRunConfigSource
)


class ScientificSelection(_SelectionModel):
    subject: SubjectChoice = Field(default_factory=UnboundSubjectChoice)
    configuration: ConfigurationChoice = Field(default_factory=ActiveConfiguration)
    batch: BatchScope = Field(default_factory=UnscopedBatch)

    def intent_content(self) -> dict[str, JsonValue]:
        return self.model_dump(mode="json")


class ReviewedScientificSelection(_SelectionModel):
    binding: ResolvedScientificBinding
    config_source: LaunchConfigSource


def require_selection_binding(
    selection: ScientificSelection, binding: ResolvedScientificBinding
) -> None:
    """Validate intent against its exact projection, without reading mutable heads."""
    subject = selection.subject
    evidence = binding.subject
    if isinstance(subject, UnboundSubjectChoice):
        if not isinstance(evidence, UnboundSubject) or isinstance(
            selection.batch, DeclaredBatch
        ):
            raise ValueError("unbound selection cannot carry sample or batch evidence")
    elif isinstance(subject, RegisteredTargetChoice):
        if (
            not isinstance(evidence, RegisteredTargetSubject)
            or evidence.ref != subject.ref
        ):
            raise ValueError("selected target differs from reviewed target")
    else:
        if not isinstance(evidence, InlineSamplesSubject) or len(evidence.samples) != 1:
            raise ValueError("sample selection requires one inline subject")
        sample = evidence.samples[0]
        if (
            sample.role != "subject"
            or sample.sample_id != subject.sample_id
            or subject.revision not in (None, sample.revision)
        ):
            raise ValueError("selected sample differs from reviewed sample")
    batch_id = (
        selection.batch.id if isinstance(selection.batch, DeclaredBatch) else None
    )
    if any(sample.batch_id != batch_id for sample in binding.samples):
        raise ValueError("selected batch differs from reviewed scientific evidence")

File: apps/scopecat-ui/src/features/launch/scientific-selection.ts
import type { LaunchPreview } from "./launch-api";
import type { components } from "../../api-schema";
import type { ConfigContextResolution } from "../config/config-api";

export type ScientificSelection = Required<components["schemas"]["ScientificSelection-Input"]>;
export function normalizeSelection(
  selection?: components["schemas"]["ScientificSelection-Input"],
): ScientificSelection {
  return {
    subject: selection?.subject ?? { kind: "unbound" },
    configuration: selection?.configuration ?? { kind: "active" },
    batch: selection?.batch ?? { kind: "unscoped" },
  };
}
export type ScientificBinding = LaunchPreview["reviewed"]["binding"];
export const defaultSelection = (): ScientificSelection => ({
  subject: { kind: "unbound" },
  configuration: { kind: "active" },
  batch: { kind: "unscoped" },
});

export function contextSelection(resolved: ConfigContextResolution): ScientificSelection {
  const source = resolved.config_source;
  return {
    subject: {
      kind: "sample",
      sample_id: source.sample.sample_id,
      revision: source.sample.revision,
    },
    configuration: { kind: "working_point", ref: source.context, overrides: source.overrides },
    batch: source.sample.batch_id
      ? { kind: "declared", id: source.sample.batch_id }
      : { kind: "unscoped" },
  };
}

export function selectedBatch(
  selection: components["schemas"]["ScientificSelection-Input"],
): string | undefined {
  return selection.batch?.kind === "declared" ? selection.batch.id : undefined;
}

export function subjectLabel(
  selection?: components["schemas"]["ScientificSelection-Input"],
): string {
  const subject = normalizeSelection(selection).subject;
  if (subject.kind === "registered_target")
    return `Registered target ${subject.ref.target_id}, revision ${subject.ref.revision}`;
  if (subject.kind === "sample")
    return `Sample ${subject.sample_id}${subject.revision ? `, revision ${subject.revision}` : " (current revision)"}`;
  return "No sample";
}

export function subjectSample(binding?: ScientificBinding) {
  const subject = binding?.subject;
  return subject?.kind === "registered_target"
    ? subject.sample
    : subject?.kind === "inline_samples"
      ? subject.samples[0]
      : undefined;
}

// openapi-fetch's Readable projection widens tuple arrays. Validate the wire
// endpoints before returning the exact request type; never discard target evidence.
export function reviewedForRequest(
  reviewed: LaunchPreview["reviewed"],
): components["schemas"]["ReviewedScientificSelection-Input"] {
  const subject = reviewed.binding.subject;
  const binding: components["schemas"]["ResolvedScientificBinding"] = {
    ...reviewed.binding,
    subject:
      subject.kind === "registered_target"
        ? {
            ...subject,
            content: {
              ...subject.content,
              connections: subject.content.connections.map((connection) => {
                if (connection.endpoints.length !== 2)
                  throw new Error("The daemon returned a target connection without two endpoints.");
                return {
                  ...connection,
                  endpoints: [connection.endpoints[0]!, connection.endpoints[1]!],
                };
              }),
            },
          }
        : subject,
  };
  return { ...reviewed, binding };
}
